api/agent/protocol/http.go

```diff
@@ -26,123 +23,40 @@ func (p *HTTPProtocol) IsStreamable() bool { return true }
-		if err != nil {
-			return err
-		}
-		res.Write(w) // TODO timeout
-	}
-
-	return nil
-}
-
-// DumpRequestTo is httputil.DumpRequest with some modifications. It will
-// dump the request to the provided io.Writer with the body always, consuming
-// the body in the process.
-//
-// TODO we should support h2!
-func DumpRequestTo(w io.Writer, req *http.Request) error {
-	// By default, print out the unmodified req.RequestURI, which
-	// is always set for incoming server requests. But because we
-	// previously used req.URL.RequestURI and the docs weren't
-	// always so clear about when to use DumpRequest vs
-	// DumpRequestOut, fall back to the old way if the caller
-	// provides a non-server Request.
-
-	reqURI := req.RequestURI
-	if reqURI == "" {
-		reqURI = req.URL.RequestURI()
-	}
-
-	fmt.Fprintf(w, "%s %s HTTP/%d.%d\r\n", valueOrDefault(req.Method, "GET"),
-		reqURI, req.ProtoMajor, req.ProtoMinor)
-
-	absRequestURI := strings.HasPrefix(req.RequestURI, "http://") || strings.HasPrefix(req.RequestURI, "https://")
-	if !absRequestURI {
-		host := req.Host
-		if host == "" && req.URL != nil {
-			host = req.URL.Host
-		}
-
-		if host != "" {
-			fmt.Fprintf(w, "Host: %s\r\n", host)
-		}
-	}
-
-	chunked := len(req.TransferEncoding) > 0 && req.TransferEncoding[0] == "chunked"
-
-	if len(req.TransferEncoding) > 0 {
-		fmt.Fprintf(w, "Transfer-Encoding: %s\r\n", strings.Join(req.TransferEncoding, ","))
-	}
-
-	if req.Close {
-		fmt.Fprintf(w, "Connection: close\r\n")
-	}
-
-	err := req.Header.WriteSubset(w, reqWriteExcludeHeaderDump)
+	resp, err := http.ReadResponse(bufio.NewReader(h.out), ci.Request()) // TODO timeout
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 
-	io.WriteString(w, "\r\n")
+	rw, ok := w.(http.ResponseWriter)
+	if !ok {
+		// async / [some] tests go through here. write a full http request to the writer
+		resp.Write(w) // TODO timeout
+		return nil
+	}
 
-	if req.Body != nil {
-		var dest io.Writer = w
-		if chunked {
-			dest = httputil.NewChunkedWriter(dest)
-		}
+	// if we're writing directly to the response writer, we need to set headers
+	// and status code, and only copy the body. resp.Write would copy a full
+	// http request into the response body (not what we want).
 
-		// TODO copy w/ ctx
-		_, err = io.Copy(dest, req.Body)
-		if chunked {
-			dest.(io.Closer).Close()
-			io.WriteString(w, "\r\n")
+	// add resp's on top of any specified on the route [on rw]
+	for k, vs := range resp.Header {
+		for _, v := range vs {
+			rw.Header().Add(k, v)
 		}
 	}
-
-	return err
-}
-
-var reqWriteExcludeHeaderDump = map[string]bool{
-	"Host":              true, // not in Header map anyway
-	"Transfer-Encoding": true,
-	"Trailer":           true,
-}
-
-// Return value if nonempty, def otherwise.
-func valueOrDefault(value, def string) string {
-	if value != "" {
-		return value
+	if resp.StatusCode > 0 {
+		rw.WriteHeader(resp.StatusCode)
 	}
-	return def
+	io.Copy(rw, resp.Body)
```

Comment:
Also for the previous code (and json side too): we are not checking error cases from Write(), WriteHeader() and io.Copy().